src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -545,15 +594,16 @@ private IEnumerable<MethodBase> FunctionLookup(IList args)
             return mbs;
         }
 
-        private static HashSet<Type> GetTypeStatisticsForArray(ExprListNode array)
+        private HashSet<Type> GetTypeStatisticsForArray(ExprListNode array, ref int rank)
```

Comment:
do you really want a `ref` type - wouldn't `out` suffice? I think you are only passing data out the method, not in both directions?

Reply:
hmm, maybe I am wrong, I missed `rank += 1;`